Asset hub: registering a foreign asset (#26)
papermoonio/polkadot-ecosystem-docs-draft: docs/build-on-polkadot/parachains/asset-hub/register-a-foreign-asset.md

- [Polkadot Asset Hub](https://polkadot.js.org/apps/?rpc=wss%3A%2F%2Fasset-hub-polkadot-rpc.dwellir.com#/explorer){target=\_blank}
- [Kusama Asset Hub](https://polkadot.js.org/apps/?rpc=wss%3A%2F%2Fsys.ibp.network%2Fstatemine#/explorer){target=\_blank}

For testing purposes, you can also interact with the Asset Hub instance of test networks:

- [Rococo Asset Hub](https://polkadot.js.org/apps/?rpc=wss%3A%2F%2Fasset-hub-rococo-rpc.dwellir.com#/explorer){target=\_blank}
- [Paseo Asset Hub](https://polkadot.js.org/apps/?rpc=wss%3A%2F%2Fpas-rpc.stakeworld.io%2Fassethub#/explorer){target=\_blank}

Before you start, ensure that you have: 

- Access to the Polkadot.js App interface, and you are connected to the desired chain
- A parachain that supports the XCMP protocol to interact with the Asset Hub parachain
- A funded wallet to pay for the transaction fees and subsequent registration of the foreign asset

> This guide will focus on using Polkadot, its local Asset Hub instance, and the [Astar](https://astar.network/){target=\_blank} parachain (`ID` 2006), as stated in the [Test Environment Setup](#test-environment-setup) section. However, the process is the same for other relay chains and their respective Asset Hub parachains, regardless of the network you are using and the parachain owner of the foreign asset.

Author: @eshaben, could you please clarify what you mean by "remove" here? I'm not sure what I should take away to avoid confusion. ty!

## Steps to Register a Foreign Asset

> 1. Open the [Polkadot.js Apps](https://polkadot.js.org/apps/){target=\_blank} interface and connect to the `Asset Hub` parachain

Reviewer: Same comments here as the local assets PR, please put UI elements in bold instead of backticks

      - For the local environment, connect to Local Node (Chopsticks), check the [Test Environment Setup](#test-environment-setup) section for more details
      - For the live networks, connect to the `Asset Hub` parachain. You can use either Polkadot, Kusama, Rococo, or Paseo Asset Hub

> 2. Click on the **Developer** tab on the left sidebar and select the **Extrinsics** section

Reviewer: I recommended a suggestion for changing this on the Local PR, I'd just transfer that over to here too. the developer tab is on the top menu bar

>     ![Access to Developer Extrinsics section](/polkadot-ecosystem-docs-draft/images/building-on-polkadot/parachains/asset-hub/register-a-foreign-asset/register-a-foreign-asset-1.webp)

Reviewer: When using steps in images, they should map to steps in the tutorial. So for example here you don't have something like this:

1. Click on **Developer**
2. Select **Extrinsics**

Therefore, there shouldn't be a 1 and 2 in the image. You can either just point to the Extrinsics item or do something like this:

```
2. Navigate to the **Extrinsics** page
   1. Click on the **Developer** tab from the top navigation bar
   2. Select **Extrinsics** from the dropdown
```

This will turn the steps into a and b though, so ideally you use "a" and "b" in the steps on the screenshot instead of numbers. There were some issues with not being able to center the letters in the step circles, Luca said he was gonna try and play around with it - not sure if he has had time to do so, but maybe you could try it out?

3. Select the **foreignAssets** pallet from the dropdown list and choose the **create** extrinsic

>     ![Select the Foreign Asset pallet](/polkadot-ecosystem-docs-draft/images/building-on-polkadot/parachains/asset-hub/register-a-foreign-asset/register-a-foreign-asset-2.webp)

Reviewer: Same comment here as above

4. Fill out the required fields and click on the copy icon to copy the **encoded call data** to your clipboard. The fields to be filled are:

    - **id** - as this is a foreign asset, the ID will be represented by a Multilocation that reflects its origin. For this case, the Multilocation of the asset will be from the source parachain perspective:
  
        ```javascript
        MultiLocation {parents: 1, interior: X1(Parachain(2006))};
        ```

    - **admin** - refers to the account that will be the admin of this asset. This account will be able to manage the asset, including updating its metadata. As the registered asset corresponds to a native asset of the source parachain, the admin account should be the sovereign account of the source parachain
      
        ??? note

>             The sibling account can be obtained through [Substrate Utilities](https://www.shawntabrizi.com/substrate-js-utilities/){target=\_blank}

Reviewer: ```suggestion
            The sibling account can be obtained through [Substrate Utilities](https://www.shawntabrizi.com/substrate-js-utilities/){target=\_blank}.
```

>             The sibling account can be obtained through [Substrate Utilities](https://www.shawntabrizi.com/substrate-js-utilities/){target=\_blank}

Reviewer: First mentioning of the sibling account, what is it?

>             ![Get parachain sovereign account](/polkadot-ecosystem-docs-draft/images/building-on-polkadot/parachains/asset-hub/register-a-foreign-asset/register-a-foreign-asset-4.webp)

Reviewer: There is a pink triangle on this image

>             ![Get parachain sovereign account](/polkadot-ecosystem-docs-draft/images/building-on-polkadot/parachains/asset-hub/register-a-foreign-asset/register-a-foreign-asset-4.webp)

Reviewer: This needs to be switched with the next image

>             Ensure that **Sibling** is selected and that the **Parachain ID** corresponds to the source parachain. In this case, since the guide follows the test setup stated in the [Test Environment Setup](./register-a-foreign-asset.md/#test-enviroment-setup) section, the **Parachain ID** is `2006`.

Reviewer: Ah ok, I would move this above the instructions on finding the sibling and then label the note like `??? note "Obtain the sibling account"`

    - **minBalance** - the minimum balance required to hold this asset

    ![Fill out the required fields](/polkadot-ecosystem-docs-draft/images/building-on-polkadot/parachains/asset-hub/register-a-foreign-asset/register-a-foreign-asset-3.webp)

>     ??? note

Reviewer: I think this should either be a `!!! note`, or if you want it to be expandable, add a title that describes what is in the note, so here it could be something like `??? note "Example encoded call data"`

This same thought applies to the other `??? note`s on this page

>         If you want an example of the encoded call data, you can copy the following: `0x3500010100591f007369626cd6070000000000000000000000000000000000000000000000000000a0860100000000000000000000000000`

Reviewer: Break this out to a new line and use 3 backticks so it can easily be copied

> 5. With the Polkadot.js interface connected to the parachain that will send the foreign asset to Asset Hub, navigate to the **Developer > Extrinsics** section. Create the following call, and paste the **encoded call data** copied in the previous step. After filling out the required fields, click the **Submit Transaction** button

Reviewer: I would separate this from the previous set of steps. Maybe have two ### sections one for actions on asset hub another for actions on the source parachain

    ![Register foreign asset through XCM](/polkadot-ecosystem-docs-draft/images/building-on-polkadot/parachains/asset-hub/register-a-foreign-asset/register-a-foreign-asset-5.webp)

    First, DOT will be withdrawn from the sibling account of the parachain. Then, the DOT will be used to initiate an execution. The transaction will be carried out with the origin kind as XCM, and the transaction will be the hex-encoded call for creating a foreign asset on Asset Hub for the specified parachain asset multilocation. Any surplus will be refunded, and the asset will be deposited into the sibling account.

    !!! warning
        Note that the sovereign account on the Asset Hub parachain must have a sufficient balance to cover the XCM `BuyExecution` instruction. If the account does not have enough balance, the transaction will fail.

    ???note

>         If you want to have the whole XCM call ready to be copied, go to the **Developer > Extrinsics > Decode** section and paste the following hex-encoded call data: `0x6300330003010100a10f030c000400010000070010a5d4e81300010000070010a5d4e80006030700b4f13501419ce03500010100591f007369626cd607000000000000000000000000000000000000000000000000000000000000000000000000000000000000`

Reviewer: triple backticks here too

        Ensure to replace the encoded call data with the one you copied in the previous step.

After the transaction is successfully executed, the foreign asset will be registered on the Asset Hub parachain. 

## Asset Registration Verification

> To confirm that a foreign asset has been successfully accepted and registered on the Asset Hub parachain, you can navigate to the `Network > Explorer` section of the Polkadot.js App interface. You should be able to see an event that includes the following details:

Reviewer: ```suggestion
To confirm that a foreign asset has been successfully accepted and registered on the Asset Hub parachain, you can navigate to the `Network > Explorer` section of the Polkadot.js Apps interface for Asset Hub. You should be able to see an event that includes the following details:
```

![Asset registration event](/polkadot-ecosystem-docs-draft/images/building-on-polkadot/parachains/asset-hub/register-a-foreign-asset/register-a-foreign-asset-6.webp)

In the image above, the **success** field indicates whether the asset registration was successful.

## Test Environment Setup

> To test the foreign asset registration process before deploying it on the live network, you can set up a local parachain environment. This guide uses `chopsticks` to simulate that process. For more information on using chopsticks, please refer to the [Chopsticks documentation](../../../dev-tools/chopsticks/overview.md){target=\_blank}.

Reviewer: ```suggestion
To test the foreign asset registration process before deploying it on a live network, you can set up a local parachain environment. This guide uses Chopsticks to simulate that process. For more information on using Chopsticks, please refer to the [Chopsticks documentation](../../../dev-tools/chopsticks/overview.md){target=\_blank}.
```

Reviewer: We shouldn't be using relative links. Not sure how many other pages are using relative links, but we'll need to address this at some point

---
title: Register a foreign asset
description: An in-depth guide to registering a foreign asset on the Asset Hub parachain, providing clear, step-by-step instructions.
---

> # Register a foreign asset on Asset Hub

Reviewer 2: ```suggestion
# Register a Foreign Asset on Asset Hub
```

## Introduction

> As outlined in the [Asset Hub Overview](./overview.md#foreign-assets){target=\_blank}, Asset Hub supports two categories of assets: local and foreign. Foreign assets originated outside of Asset Hub and are recognized by [Multilocations](https://wiki.polkadot.network/docs/learn/xcm/fundamentals/multilocation-summary){target=\_blank}. This guide will walk you through the process of registering a foreign asset on the Asset Hub parachain.

Reviewer 2: I think that we don't use `{target=\_blank}` for links to local pages

Author: Mm as far as I understand, we don't use `{target=\_blank}` for links to the same page, as here I'm referring to the page `overview.md` from `register-a-foreign-asset.md`, then I guess it should be included.

Here is the reference: https://github.com/papermoonio/documentation-style-guide/blob/main/style-guide.md#links

> As outlined in the [Asset Hub Overview](./overview.md#foreign-assets){target=\_blank}, Asset Hub supports two categories of assets: local and foreign. Foreign assets originated outside of Asset Hub and are recognized by [Multilocations](https://wiki.polkadot.network/docs/learn/xcm/fundamentals/multilocation-summary){target=\_blank}. This guide will walk you through the process of registering a foreign asset on the Asset Hub parachain.

Reviewer 2: ```suggestion
As outlined in the [Asset Hub Overview](./overview.md#foreign-assets){target=\_blank}, Asset Hub supports two categories of assets: local and foreign. Foreign assets originated outside of Asset Hub are recognized by [Multilocations](https://wiki.polkadot.network/docs/learn/xcm/fundamentals/multilocation-summary){target=\_blank}. This guide will walk you through the process of registering a foreign asset on the Asset Hub parachain.
```

- [Polkadot Asset Hub](https://polkadot.js.org/apps/?rpc=wss%3A%2F%2Fasset-hub-polkadot-rpc.dwellir.com#/explorer){target=\_blank}
- [Kusama Asset Hub](https://polkadot.js.org/apps/?rpc=wss%3A%2F%2Fsys.ibp.network%2Fstatemine#/explorer){target=\_blank}

For testing purposes, you can also interact with the Asset Hub instance of test networks:

- [Rococo Asset Hub](https://polkadot.js.org/apps/?rpc=wss%3A%2F%2Fasset-hub-rococo-rpc.dwellir.com#/explorer){target=\_blank}
- [Paseo Asset Hub](https://polkadot.js.org/apps/?rpc=wss%3A%2F%2Fpas-rpc.stakeworld.io%2Fassethub#/explorer){target=\_blank}

Before you start, ensure that you have: 

- Access to the Polkadot.js App interface, and you are connected to the desired chain
- A parachain that supports the XCMP protocol to interact with the Asset Hub parachain
- A funded wallet to pay for the transaction fees and subsequent registration of the foreign asset

> This guide will focus on using Polkadot, its local Asset Hub instance, and the [Astar](https://astar.network/){taget=\_blank} parachain (`ID` 2006), as stated in the [Test Environment Setup](./register-a-foreign-asset.md/#test-enviroment-setup) section. However, the process is the same for other relay chains and their respective Asset Hub parachains, regardless of the network you are using and the parachain owner of the foreign asset.

Reviewer 2: ```suggestion
This guide will focus on using Polkadot, its local Asset Hub instance, and the [Astar](https://astar.network/){target=\_blank} parachain (`ID` 2006), as stated in the [Test Environment Setup](#test-environment-setup) section. However, the process is the same for other relay chains and their respective Asset Hub parachains, regardless of the network you are using and the parachain owner of the foreign asset.
```

## Steps to Register a Foreign Asset

1. Open the [Polkadot.js Apps](https://polkadot.js.org/apps/){target=\_blank} interface and connect to the `Asset Hub` parachain

>       - For the local environment, connect to Local Node (Chopsticks), check the [Test Environment Setup](./register-a-foreign-asset.md/#test-enviroment-setup) section for more details

Reviewer 2: ```suggestion
      - For the local environment, connect to Local Node (Chopsticks), check the [Test Environment Setup](#test-environment-setup) section for more details
```

      - For the live networks, connect to the `Asset Hub` parachain. You can use either Polkadot, Kusama, Rococo, or Paseo Asset Hub

2. Click on the **Developer** tab on the left sidebar and select the **Extrinsics** section

    ![Access to Developer Extrinsics section](/polkadot-ecosystem-docs-draft/images/building-on-polkadot/parachains/asset-hub/register-a-foreign-asset/register-a-foreign-asset-1.webp)

> 3. Select the **Foreign Asset** pallet from the dropdown list and choose the **create** extrinsic

Reviewer 2: ```suggestion
3. Select the **foreignAssets** pallet from the dropdown list and choose the **create** extrinsic
```

>     ![Select the Foreign Asset pallet](/polkadot-ecosystem-docs-draft/images/building-on-polkadot/parachains/asset-hub/register-a-foreign-asset/register-a-foreign-asset-2.webp)

Reviewer 2: The arrows in this image are smaller than in other images

Author: Will re-take all screenshots according to the discussion we had yesterday

> 4. Fill out the required fields and click on the copy icon to copy the **encoded call** to your clipboard

Reviewer 2: ```suggestion
4. Fill out the required fields and click on the copy icon to copy the **encoded call data** to your clipboard
```

>     ![Fill out the required fields](/polkadot-ecosystem-docs-draft/images/building-on-polkadot/parachains/asset-hub/register-a-foreign-asset/register-a-foreign-asset-3.webp)

Reviewer 2: The arrows in this image are smaller than in other images

    ??? note
        If you want an encoded call example, you can copy the following: `0x3500010100591f007369626cd6070000000000000000000000000000000000000000000000000000a0860100000000000000000000000000`

>     The fields to be filled are:

Reviewer 2: I would move this above the image

    - **id** - as this is a foreign asset, the ID will be represented by a Multilocation that reflects its origin. For this case, the Multilocation of the asset will be from the source parachain perspective:
  
        ```javascript
        MultiLocation {parents: 1, interior: X1(Parachain(2006))};
        ```

    - **admin** - refers to the account that will be the admin of this asset. This account will be able to manage the asset, including updating its metadata. As the registered asset corresponds to a native asset of the source parachain, the admin account should be the sovereign account of the source parachain. 
      
        ??? note

>             The sibling account can be obtained through the [Substrate JS set of utilities](https://www.shawntabrizi.com/substrate-js-utilities/){target=\_blank}

Reviewer 2: ```suggestion
            The sibling account can be obtained through [Substrate Utilities](https://www.shawntabrizi.com/substrate-js-utilities/){target=\_blank}
```

            ![Get parachain sovereign account](/polkadot-ecosystem-docs-draft/images/building-on-polkadot/parachains/asset-hub/register-a-foreign-asset/register-a-foreign-asset-4.webp)

            Ensure that **Sibling** is selected and that the **Parachain ID** corresponds to the source parachain. In this case, since the guide follows the test setup stated in the [Test Environment Setup](./register-a-foreign-asset.md/#test-enviroment-setup) section, the **Parachain ID** is `2006`.

    - **minBalance** - the minimum balance required to hold this asset

> 5. With the Polkadot.js interface connected to the parachain that will send the foreign asset to Asset Hub, navigate to the **Developer > Extrinsics** section. Create the following call, and paste the **encoded call** copied in the previous step. After filling out the required fields, click the **Submit Transaction** button.

Reviewer 2: ```suggestion
5. With the Polkadot.js interface connected to the parachain that will send the foreign asset to Asset Hub, navigate to the **Developer > Extrinsics** section. Create the following call, and paste the encoded call data copied in the previous step. After filling out the required fields, click the **Submit Transaction** button.
```

    ![Register foreign asset through XCM](/polkadot-ecosystem-docs-draft/images/building-on-polkadot/parachains/asset-hub/register-a-foreign-asset/register-a-foreign-asset-5.webp)

    First, DOT will be withdrawn from the sibling account of the parachain. Then, the DOT will be used to initiate an execution. The transaction will be carried out with the origin kind as XCM, and the transaction will be the hex-encoded call for creating a foreign asset on Asset Hub for the specified parachain asset multilocation. Any surplus will be refunded, and the asset will be deposited into the sibling account.

    !!! warning
        Note that the sovereign account on the Asset Hub parachain must have a sufficient balance to cover the XCM `BuyExecution` instruction. If the account does not have enough balance, the transaction will fail.

    ???note

>         If you want to have the whole XCM call ready to be copied, go to go to the **Developer > Extrinsics > Decode** section and paste the following hex-encoded call: `0x6300330003010100a10f030c000400010000070010a5d4e81300010000070010a5d4e80006030700b4f13501419ce03500010100591f007369626cd607000000000000000000000000000000000000000000000000000000000000000000000000000000000000`

Reviewer 2: ```suggestion
        If you want to have the whole XCM call ready to be copied, go to the **Developer > Extrinsics > Decode** section and paste the following hex-encoded call: `0x6300330003010100a10f030c000400010000070010a5d4e81300010000070010a5d4e80006030700b4f13501419ce03500010100591f007369626cd607000000000000000000000000000000000000000000000000000000000000000000000000000000000000`
```

        Ensure to replace the encoded call with the one you copied in the previous step.

After the transaction is successfully executed, the foreign asset will be registered on the Asset Hub parachain. 

## Asset Registration Verification

> To confirm that a foreign asset has been successfully accepted and registered on the Asset Hub parachain, you can navigate to the `Network > Explorer` section of the Polkadot.Js App interface. You should be able to see an event that includes the following details:

Reviewer 2: ```suggestion
To confirm that a foreign asset has been successfully accepted and registered on the Asset Hub parachain, you can navigate to the `Network > Explorer` section of the Polkadot.js App interface. You should be able to see an event that includes the following details:
```

- [Paseo Asset Hub](https://polkadot.js.org/apps/?rpc=wss%3A%2F%2Fpas-rpc.stakeworld.io%2Fassethub#/explorer){target=\_blank}

Before you start, ensure that you have: 

- Access to the Polkadot.js Apps interface, and you are connected to the desired chain
- A parachain that supports the XCMP protocol to interact with the Asset Hub parachain
- A funded wallet to pay for the transaction fees and subsequent registration of the foreign asset

This guide will use Polkadot, its local Asset Hub instance, and the [Astar](https://astar.network/){target=\_blank} parachain (`ID` 2006), as stated in the [Test Environment Setup](#test-environment-setup) section. However, the process is the same for other relay chains and their respective Asset Hub parachain, regardless of the network you are using and the parachain owner of the foreign asset.

## Steps to Register a Foreign Asset

### Asset Hub

1. Open the [Polkadot.js Apps](https://polkadot.js.org/apps/){target=\_blank} interface and connect to the Asset Hub parachain using the network selector in the top left corner 

>       - Testing foreign asset registration is recommended on TestNet before proceeding to MainNet. If you haven't set up a local testing environment yet, consult the [Environment setup](#environment-setup) guide. After setting up, connect to the Local Node (Chopsticks) at `ws://127.0.0.1:8000`

Reviewer: ```suggestion
      - Testing foreign asset registration is recommended on TestNet before proceeding to MainNet. If you haven't set up a local testing environment yet, consult the [Environment setup](#test-environment-setup) guide. After setting up, connect to the Local Node (Chopsticks) at `ws://127.0.0.1:8000`
```

      4. Choose the **create** extrinsic

    ![Select the Foreign Asset pallet](/polkadot-ecosystem-docs-draft/images/building-on-polkadot/parachains/asset-hub/register-a-foreign-asset/register-a-foreign-asset-2.webp)

3. Fill out the required fields and click on the copy icon to copy the **encoded call data** to your clipboard. The fields to be filled are:

    - **id** - as this is a foreign asset, the ID will be represented by a Multilocation that reflects its origin. For this case, the Multilocation of the asset will be from the source parachain perspective:
  
        ```javascript
        MultiLocation {parents: 1, interior: X1(Parachain(2006))};
        ```

    - **admin** - refers to the account that will be the admin of this asset. This account will be able to manage the asset, including updating its metadata. As the registered asset corresponds to a native asset of the source parachain, the admin account should be the sovereign account of the source parachain
      
        ??? note "Obtain the sovereign account"
            The sovereign account can be obtained through [Substrate Utilities](https://www.shawntabrizi.com/substrate-js-utilities/){target=\_blank}.

>             Ensure that **Sibling** is selected and that the **Para ID** corresponds to the source parachain. In this case, since the guide follows the test setup stated in the [Test Environment Setup](#test-enviroment-setup) section, the **Para ID** is `2006`.

Reviewer: ```suggestion
            Ensure that **Sibling** is selected and that the **Para ID** corresponds to the source parachain. In this case, since the guide follows the test setup stated in the [Test Environment Setup](#test-environment-setup) section, the **Para ID** is `2006`.
```

>             Ensure that **Sibling** is selected and that the **Para ID** corresponds to the source parachain. In this case, since the guide follows the test setup stated in the [Test Environment Setup](./register-a-foreign-asset.md/#test-enviroment-setup) section, the **Para ID** is `2006`.

Reviewer: ```suggestion
            Ensure that **Sibling** is selected and that the **Para ID** corresponds to the source parachain. In this case, since the guide follows the test setup stated in the [Test Environment Setup](#test-enviroment-setup) section, the **Para ID** is `2006`.
```
